### PR TITLE
Swapper Implant

### PR DESCRIPTION
**Pull request by FluffTheMoth01**
## About The Pull Request
This PR adds the swapper implant to the game as an 5 TC traitor item. This implant allows a smart traitor to stealthily inject somebody with an implant, and then at a later time use an activator pen to swap places with the mob the implant was injected in.

This is a one time use item, so it needs to be used wisely.

## Why It's Good For The Game
This allows a smart traitor a lot of options, they can use it to break into secure areas, to escape from the police, or perhaps even frame somebody for a crime they didn't commit. This is much more inline to the kind of traitor items I like to see, which are weird quirky gadgets that don't just work in a straightforward manner.

Using this is a big risk as you don't know exactly where the implant is any any given point in time, but it could also be incredibly effective if used correctly.

## Changelog
:cl:
add: The Swapper Implant is now in traitor uplinks for 5 TC
/:cl:

### CODE REVIEW

**Comment by XDTM** [comment]
It's worth noting this can be already achieved by cavity implanting a swapper into someone's chest, although it lacks the whole stealth aspect.

File: code/game/objects/items/implants/implantswap.dm [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+		RL.visible_message("[RL] Suddenly vanishes, Leaving [AL] in their place!")
+		AL.visible_message("[AL] Suddenly vanishes, leaving [RL] in their place!")
+		to_chat(RL, "You suddenly find yourself in a new location.")
+		to_chat(AL, "You press the button on the pen, and suddenly find yourself in a new location.")
+
+		///Move the reciever to the activator
+		RL.forceMove(AT)
+		///Move the activator to the reciever
+		AL.forceMove(RT)
```
Comment (XDTM) [inline code/game/objects/items/implants/implantswap.dm]: Use the do_teleport proc instead of forceMove, ideally with a TELEPORT_CHANNEL_QUANTUM since it's a couple of paired devices.

File: code/game/objects/items/implants/implantswap.dm [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+/obj/item/pen/swap_activator/attack_self()
+	swap.activate()
+
+/obj/item/implant/swapper
+	var/datum/swap/swap
+
+/obj/item/implant/swapper/implant(silent = TRUE, force = TRUE)
+	swap.implanted = TRUE
+	. = ..()
```
Comment (nemvar) [inline code/game/objects/items/implants/implantswap.dm]: Replace with
```
/obj/item/implant/swapper/implant(silent = TRUE, force = TRUE)
	. = ..()
	if(.)
		swap.implanted = TRUE

File: code/game/objects/items/implants/implantswap.dm [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+		to_chat(AL, "You press the button on the pen, and suddenly find yourself in a new location.")
+
+		///Move the reciever to the activator
+		RL.forceMove(AT)
+		///Move the activator to the reciever
+		AL.forceMove(RT)
+
+		///Cleans up the objects after use.
+		qdel(R)
+		qdel(A)
```
Comment (nemvar) [inline code/game/objects/items/implants/implantswap.dm]: QDEL_NULL

File: code/game/objects/items/implants/implantswap.dm [hunk 4]
```diff
@@ -0,0 +1,65 @@
+/datum/swap
```
Comment (nemvar) [inline code/game/objects/items/implants/implantswap.dm]: This entire datum isn't needed. It's only used to hold two variables and a proc. Either turn it into a full scale component or remove it.

**Comment by Dawson1917** [comment]
8tc feels like a bit much for a one-use, risky item. If you're swapping into a secure area (vault, ai upload, armory) you're going to be boned unless you also have access to get out, and if you did then why not just walk inside in the first place? 6tc or even 5tc might be better considering 8tc is on the level of adrenals, emags and eswords.

**Comment by nemvar** [comment]
The real value of this is that it can basically be used to kill anyone who has been implanted. But because the syringe doesn't appear to inject instantly, this will probably be in the trash tier for traitor items and only be used by memesters, noobs and masochists.

**Comment by SpacePrius** [comment]
@nemvar does it not inject instantly? I thought that was what the forced variable did, let me modify that

**Comment by Dawson1917** [comment]
> 
> 
> The real value of this is that it can basically be used to kill anyone who has been implanted. But because the syringe doesn't appear to inject instantly, this will probably be in the trash tier for traitor items and only be used by memesters, noobs and masochists.

mfw implant cap, grab eva suit and metal, build 4 walls around myself in space, swap places and come back to loot his corpse
You're right, best use would probably be to kill someone out of nowhere. Needs to inject instantly since TG players have a heart attack whenever someone stands next to them for more than a microsecond

**Comment by nemvar** [comment]
No idea if this injects instantly, I assumed the forced variable would make it ignore armor, but could be wrong about that. This is pretty nutty if it is indeed instant.

**Comment by SpacePrius** [comment]
alright now it injects instantly, didn't catch that originally

**Comment by Shadowflame909** [comment]
Damn. This combined with xhuis's cremation PR would be one nifty PR.

Inject yourself with it, inject your target that you gotta kill. Load up the xenobio gas chamber. Heat up that plasma, then run into it and quickly use the swapper.

Come back later to a completely cremated corpse.

**Comment by wesoda25** [comment]
Why do you keep calling XDTM xhuis

**Comment by Shadowflame909** [comment]
Thanks for the correction nervere.

**Comment by anconfuzedrock** [comment]
while one use this feels kinda bs to me 
how to be lame with this item: 
1. inject 4 boys with this
2. space yourself in a hardsuit or even just a firesuit
swap with person to have them die of space
3. repeat
4. for foolproofness just make sure at least one of them is somewhere safe then you can swap to them last

**Comment by Shadowflame909** [comment]
I agree with your points Anconfusedrock. But I feel like this item can offer a whole new amount of creative deaths/kill-chambers. Think like saw.

Essentially what I mean is, before it was just e-sword/e-bow as a traitor. Holoparasite, C4 + Bola. The same old Loud kill you before you can react kind of stuff.

This offers something similar to that in purpose. Except the traitor won't be directly killing you. Their imaginative kill room will. 

Lets say you're teleported to a plasma-flood room randomly. You're going to be like "what the fuck-" but as this item becomes more common. Sooner or later you're going to be instantly preparing the comms ;SOMEONE CHECK MY SUIT SENSORS IM BURNING UP IN A PLASMA KILL ROOM.

Maybe the death will be even faster than that. Like a bomb rigged to emitters or something in a 5x5 closed room in space. 

Well then this can be addressed by a simple line of text, that's probably already in existence. Maybe just rip some abductor code and have the warp leave you disoriented so someone looks around before running into any death traps.

Tldr; Good idea. Much better then what we already have. A creative traitor is the best traitor.

**Comment by Dawson1917** [comment]
There's a large lack of creativity in how people are killed because you can alert the entire station in half a second by yelling ";HELP", suit sensors can show your exact position in real-time to every doctor with a crew monitor in their pocket, the AI can jump to you instantly, hallways are always filled, and maintenance always has at least three people running through it at any given moment. Resulting in most "stealthy" traitor items boiling down to some kind of stun (sleepy pen) or surprise attack (stealth box, cham projector) so you can beat someone to death and run. Even though there's easymode potential for cheesing kills with this, and that people are going to be salty over dying to something they can't counter in any way other than barricading themselves off from the rest of the station at roundstart, being able to move people to an exact position might make for actually interesting situations both the antag and their victim can get a kick out of.

**Comment by zxaber** [comment]
Seems more like a neat "get out of jail free" card. Inject someone who you're sure will stay on the other side of the station (or inject an empty clone lmao), and do your murder bone. When you get backed into a corner, just swap, change outfit, and continue on like you weren't the mass-murderer.

Presumably, you'll get the "you feel a prick" message, the syringe will have an in-hand sprite, and they'll have to be within one tile. As such, I wouldn't consider it really _that_ unfair to die to; you get plenty of warnings. It's really no worse than the sleepy pen.

File: code/modules/uplink/uplink_items.dm [hunk 5]
```diff
@@ -1567,6 +1567,13 @@ GLOBAL_LIST_INIT(uplink_items, subtypesof(/datum/uplink_item))
 	surplus = 0
 	restricted = TRUE
 
+	/datum/uplink_item/implants/swapper
+	name = "Swapper Implant"
+	desc = "The Swapper Implant comes in two parts, an activator pen, and an implant. The implant can be stealthily injected into any living creature, and can remain inside them indefinetly. \
```
Comment (bgobandit) [inline code/modules/uplink/uplink_items.dm]: indefinitely

File: code/game/objects/items/implants/implantswap.dm [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+	if(R.implanted && R && A)
+		///Grab the recievers turf
+		var/turf/RT = get_turf(RL.loc)
+		///Grab the activators turf
+		var/turf/AT = get_turf(A.loc)
+		
+		playsound(RL, 'sound/effects/sparks4.ogg')
+		playsound(AL, 'sound/effects/sparks4.ogg')
+
+		RL.visible_message("[RL] suddenly vanishes, leaving [AL] in their place!")
```
Comment (bgobandit) [inline code/game/objects/items/implants/implantswap.dm]: p_their

File: code/game/objects/items/implants/implantswap.dm [hunk 7]
```diff
@@ -0,0 +1,74 @@
+/*!
+The swapper implant comes in two parts, the activator pen, and the swapper implant itself. The activator stores the object reference to the implant, and uses that to swap the locations of the mobs holding the objects.area
+*/
+
+/obj/item/pen/swap_activator
+	///Stores the implant object ref.
+	var/obj/item/implant/swapper/reciever
```
Comment (bgobandit) [inline code/game/objects/items/implants/implantswap.dm]: misspellings in vars makes the english language cry

**Comment by bgobandit** [comment]
I like this and agree with you about the sorts of traitor items we need more/less of, but can it have some more fluff text about stolen bluespace technology or whatever

**Comment by wesoda25** [comment]
This is honestly really wack. I don’t get it or why we need it. Not to mention while its usage is pretty niche, it seems pretty OP.

File: code/game/objects/items/implants/implantswap.dm (hunk 6 above)
Comment (SpacePrius) [reply code/game/objects/items/implants/implantswap.dm]: Their is spelled correctly.

**Comment by PKPenguin321** [comment]
New parapen C4, inject someone with one of these then shoot yourself out of a mass driver and swap with em. I'll give that it does give them a chance to yell about it assuming the injection isn't silent, but there's basically nothing they can reasonably do to save themselves once they've been hit with the tiny prick.

**Comment by SkeletalElite** [comment]
Cool idea but Im not a huge fan of adding a traitor item that lets you be teleported into space suddenly without anyway to avoid other than be the most paranoid spaceman to ever exist other than run away and become a manit hobo who beats anyone entering his maint cave.

**Comment by Whoneedspacee** [comment]
why do you tempt me

File: code/game/objects/items/implants/implantswap.dm [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+The swapper implant comes in two parts, the activator pen, and the swapper implant itself. The activator stores the object reference to the implant, and uses that to swap the locations of the mobs holding the objects.area
+*/
+
+/obj/item/pen/swap_activator
+	///Stores the implant object ref.
+	var/obj/item/implant/swapper/receiver
+
+/obj/item/pen/swap_activator/proc/activate(var/obj/item/implant/swapper/R = receiver, var/obj/A = src)
+	///Grab the recievers mob
+	var/mob/living/RL = R.loc
```
Comment (Whoneedspacee) [inline code/game/objects/items/implants/implantswap.dm]: Please no simplemobs, I already see this being used to just bring megafauna and other things to the station. Also it being a simplemob really does defeat the purpose of the item in the first place (to be in a place you're not sure what it is).

File: code/game/objects/items/implants/implantswap.dm [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+	if(R.implanted && R && A)
+		///Grab the recievers turf
+		var/turf/RT = get_turf(RL.loc)
+		///Grab the activators turf
+		var/turf/AT = get_turf(A.loc)
+		
+		playsound(RL, 'sound/effects/sparks4.ogg')
+		playsound(AL, 'sound/effects/sparks4.ogg')
+
+		RL.visible_message("[RL] suddenly vanishes, leaving [AL] in their place!")
```
Comment (Whoneedspacee) [inline code/game/objects/items/implants/implantswap.dm]: Visible messages can also send chatlogs to the user as a second argument, use that instead of just doing the to_chats below again.

**Comment by Shadowflame909** [comment]
If we're willing for this item to be nerfed. Then there ar emultiple ideas I can think off that should ease some of the parapen/c4 instant kill players fears.

1. Have the implant removable like the majority of implants in this game

2. Being Emp'd destroys the implant internally/being shocked does.

And something already in this game but my idea is. If you get implanted and get a "you feel a tiny prick.." message. You should alert literally everyone to it, and then grab a tracking implant from sec/get one of those tiny space/lavaland coordinates device that you can put your name on.

There already exists a pretty interesting way for traitors to space people. But clearly, this PR will lead to more than that. You just gradually toss a tracking beacon to the next z-level. out of an airlock. Get a hand-teleporter and then open it in the middle of a station-hallway. Someone goes through and then you close it before they get back manually.

gg.

**Comment by CRITAWAKETS** [comment]
Maybe also just make the teleportation range limited. I dont mean like 5 tiles away but something like 25 and no z-level crossing. That way you have to get your target to come close to the trap before being teleported to where you want the target to be.

**Comment by Shadowflame909** [comment]
you'd rather buy a support holoparasite and make a kill-trap that way honestly. Much less useful then the 18 tc healing and killing nerd

**Comment by TheRespectedMan** [comment]
5 TC is really cheap, you'll kill your target and still have 15 TC to spare.
"but it could also be incredibly effective if used correctly."
could never be more accurate.

Pull the pin on a syndicate minibomb in a narrow room and then swap.
Aggro a megafauna and then swap

I'm definitely against this. Just another PR that hasn't put more than 2 seconds of thought into game balance.

**Comment by TheRespectedMan** [comment]
I look forward to swapping catgirls into the incinerator for 5 tc

File: code/game/objects/items/implants/implantswap.dm (hunk 9 above)
Comment (SpacePrius) [reply code/game/objects/items/implants/implantswap.dm]: I honestly prefer to keep my functions atomic, which is why I use to_chat, also visible message doesn't support HTML spans I believe.

File: code/game/objects/items/implants/implantswap.dm (hunk 8 above)
Comment (SpacePrius) [reply code/game/objects/items/implants/implantswap.dm]: The reason i made it simplemobs is exactly that. I want the uses to be open ended and unrestricted.

Comment (Whoneedspacee) [reply code/game/objects/items/implants/implantswap.dm]: Yeah but for 5 tc that’s ridiculous how easy that would be.

You could literally get 4 megafauna on the station just by running away from them.

**Comment by SpacePrius** [comment]
You lads do realize that the implant is derived from the basic implant, it works exactly the same as ALL other implants in the game by default. There are counters to it, it just requires you knowing what you are doing.

**Comment by PKPenguin321** [comment]
if by "knowing what you're doing" you mean "immediately killing yourself and getting cloned before you get spaced" or "making a mad dash to the teleporter room to try and steal a hand tele" then sure

**Comment by Qydra** [comment]
Creative, not mass-murderey, and requires at least some effort to make big plays with it.  I approve.

**Comment by CPTANT** [comment]
I like it but would really like to see restriction on just swapping people straight into outer space.

**Comment by Dawson1917** [comment]
> 
> 
> I like it but would really like to see restriction on just swapping people straight into outer space.

The way around that would be to depressurize a room before swapping. Bash a window in the abandoned bar or abandoned medical, let all the air out and build 4 walls around you. Swap and come back to loot the corpse

**Comment by Shadowflame909** [comment]
I still don't see how that's bad. It's even more effort than filling a syringe gun up with lexorin. Buying a C4 + bola and throwing it at someone. Or just going to town with e-sword + e-bow

**Comment by CPTANT** [comment]
> 
> 
> > I like it but would really like to see restriction on just swapping people straight into outer space.
> 
> The way around that would be to depressurize a room before swapping. Bash a window in the abandoned bar or abandoned medical, let all the air out and build 4 walls around you. Swap and come back to loot the corpse

Just don't allow swapping to rooms with low/no atmosphere in general. Sure there are ways around that but it would at least require some effort.

**Comment by optimumtact** [comment]
I actually had some ideas about suit sensors only updating every couple of minutes, so you couldn't game them, but they could still be used to find dead people

**Comment by zxaber** [comment]
Isn't the point of suit sensors that it's a tradeoff between possibly being found by a medic while dead and possibly being found by a traitor while alive?

**Review by monkeylover1221 — APPROVED** [review]

**Comment by optimumtact** [comment]
@monkeylover1221 only warning, do not use the review system as a non maintainer or you will be blocked from the repo

**Comment by Akrilla** [comment]
Could you add this to the contractor random item pool? Would be a good addition

File: code/game/objects/items/implants/implantswap.dm (hunk 8 above)
Comment (Akrilla) [reply code/game/objects/items/implants/implantswap.dm]: This is a pretty massive feature to give the swappers, which could cause unintended balance issues. If it becomes a problem, it's an easy removal though.

File: code/game/objects/items/implants/implantswap.dm (hunk 9 above)
Comment (Akrilla) [reply code/game/objects/items/implants/implantswap.dm]: visible_message allows for HTML, and these probably should be span'd with notice or alternative. In general I don't think unstyled text is wanted, and something I've been asked to change before

**Comment by CRITAWAKETS** [comment]
> I still don't see how that's bad. It's even more effort than filling a syringe gun up with lexorin. Buying a C4 + bola and throwing it at someone. Or just going to town with e-sword + e-bow

The issue with what you are saying is that those methods are very obvious and can be countered, while the swapper is basically just instant death.

File: code/game/objects/items/implants/implantswap.dm (hunk 8 above)
Comment (optimumtact) [reply code/game/objects/items/implants/implantswap.dm]: Don't allow them to swap megafauna please, we already have enough trouble with them already.

File: code/game/objects/items/implants/implantswap.dm (hunk 9 above)
Comment (SpacePrius) [reply code/game/objects/items/implants/implantswap.dm]: still the point about atomic, but I will introduce the notices.

File: code/game/objects/items/storage/uplink_kits.dm [hunk 10]
```diff
@@ -267,6 +267,7 @@
 	new /obj/item/modular_computer/tablet/syndicate_contract_uplink/preset/uplink(src)
 	new /obj/item/storage/box/syndicate/contractor_loadout(src)
 	new /obj/item/melee/classic_baton/telescopic/contractor_baton(src)
+	new /obj/item/storage/box/syndie_kit/swapper_implant
```
Comment (Akrilla) [inline code/game/objects/items/storage/uplink_kits.dm]: Not here please :sweat_smile:, the `item_list` underneath. `item_list` gets randomly chosen from while here would be a permanent item.

**Comment by CPTANT** [comment]
Having thought about it I think this can be balanced quite well with the range restriction.  It allows scaling from infinite range (completely OP) to 1 tile range (useless). Somewhere there will be a range that will make this a fun and balanced tool.

**Comment by SpacePrius** [comment]
Not having it be infinite ranged ruins the downside of the mechanic, being that you don't know exactly where you are going to pop out when you do the swap. If you swap into a populated area it would be pretty obvious what you are doing.

**Comment by StonebayKyle** [comment]
I think the main reason why people are hesitant is this:
It allows for instant, uncounterable death once pricked, with no way of knowing its about to happen.

I personally really like this idea as it could lead to a lot of creative killing solutions, but in its current state, it is too easy to walk out into space and swap, which is definitely not interesting or fun.

Also, this should be balanced against freedom implants, cryptographic sequencers, and other tools that can get you out of the hands of security. 5TC may be way too cheap if it's an instant, (again) uncounterable way of getting out of permabrig or execution.

Here's a few solutions I propose:
As, anconfuzedrock suggested, make it similar to heat signature's swapper. On screen multi-use swapping. However, this could work really cool as a separate item that could be the weaker version of this item.

As for this item...
Keep the long range teleportation in, but make it so you must stand still for a very long amount of time (similar to breaking out of cuffs). This would allow security to counter the teleportation WITHOUT metagaming, as security is always moving you around to stop you from breaking out of your cuffs/buckle anyway. (possibly make the charging time proportionate to the distance between you and your victim, similar to assimilating somebody before that was removed)

This would also counter swapping people into space without the proper equipment. On top of needing a hardsuit to pull it off, you'd also need a way to keep yourself still in space, which either needs to be done by a jetpack or a lattice. The lattice would give the victim a chance to propel themselves through space guaranteeing they'd run into the station at some point to get cloned.

On the last 5-15 seconds of charging up the swap (maybe 10% if charging time varies), enable some kind of aura or something that tells the crew you're about to do something nasty, kind of like how the crew can see you succeed to break from buckles while still being restrained.

Also, disorient the person using the swapper during the last portion of charging it and after they successfully teleport, but do not disorient the swapper victim.

Final note: This item, even after all these changes, would still be too cheap as a 5TC item. I'm thinking something like 8-10TC.

Right now, without these changes, I think its way too easy to kill somebody without getting creative. I believe these changes and maybe others that people suggest could mitigate that without nerfing the item into uselessness.

Good like, SpacePrius, I'd love to see a swapper come into existence eventually.

**Comment by BlueNothing** [comment]
This item practically already exists as a *non-antag* item in the form of the Quantum Spin Inverter. Build a deathtrap or whatever, slip one in someone's backpack, go into deathtrap, fire inverter, make the other end non-viable somehow so you don't get swapped right back into it.

**Comment by Akrilla** [comment]
Giving a smaller, more accessible version of it for TC feels reasonable.

**Comment by BlueNothing** [comment]
So IMO, the TC version doesn't really need much of a nerf. If you want, I say make it, oh, 7 TC, and make it multi-use, but still single target only.

**Comment by optimumtact** [comment]
Honestly I don't like that this removes creative solutions that already exist in favour of just tc spend.
:-1: from me

**Comment by Akrilla** [comment]
I think this encourages creative solutions - if you're going for creative you're going to do it either way, having a TC item that basically screams "you can do cool stuff with this" will hopefully have traitors that weren't ever going to do something unique at least give it a shot.

**Comment by wesoda25** [comment]
It just over saturates the already over saturated uplink with something that will hardly ever be used though. Seems like a really weird item used in niche cases only. And honestly can’t wait for johhny traitor to just randomly teleport me into bubblegum. Its a cool idea but it sounds sort of like a TC trade.

**Comment by Akrilla** [comment]
keep it in as a contract kit exclusive 😎

**Comment by optimumtact** [comment]
@tgstation/commit-access closing in 24 hours unless someone merges

**Comment by StonebayKyle** [comment]
What if we testmerge so we can see if it really becomes an issue? And if so, remove it and have somebody fix it? I still like the idea of a TC-bought swapper

**Comment by SpaceManiac** [comment]
conflict

**Comment by Akrilla** [comment]
pls space we want this

**Comment by SpaceManiac** [comment]
ping a maintainer to have this reopened
